pkg/controller/clone/planner.go

```diff
@@ -0,0 +1,493 @@
+package clone
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/go-logr/logr"
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+// Planner plans clone operations
+type Planner struct {
+	RootObjectType  client.ObjectList
+	OwnershipLabel  string
+	UIDField        string
+	Image           string
+	PullPolicy      corev1.PullPolicy
+	InstallerLabels map[string]string
+	Client          client.Client
+	Recorder        record.EventRecorder
+	Controller      controller.Controller
+
+	watchingCore      bool
+	watchingSnapshots bool
+	watchMutex        sync.Mutex
+}
+
+// Phase is the interface implemeented by all clone phases
+type Phase interface {
+	Name() string
+	Reconcile(context.Context) (*reconcile.Result, error)
+}
+
+// ProgressReporter allows a phase to report progress
+type ProgressReporter interface {
+	Progress(context.Context) (string, error)
+}
+
+// list of all possible (core) types created
+var coreTypesCreated = []client.Object{
+	&corev1.PersistentVolumeClaim{},
+	&corev1.Pod{},
+}
+
+// AddCoreWatches watches "core" types
+func (p *Planner) AddCoreWatches(log logr.Logger) error {
```

Comment:
Why do we add this watch here instead of adding it to the clone populator? Why do we need to watch core types if we already have watches for PVCs? I'm missing something here.

Reply:
Yeah this is a little wonky.  The idea is that the clone populator would be ignorant of the types of resources created by the phases so the planner would handle creating the watches.  This includes an additional PVC watch because the one in base populator does not cover PVCs with ownership label.

BUT the clone populator currently knows what types are created because they are needed for the cleanup function here:  https://github.com/mhenriks/containerized-data-importer/blob/clone-pop/pkg/controller/populators/clone-populator.go#L274-L278

I think it would be better if the clone controller either 1) gets the types to cleanup from the planner or 2) calls a function on the planner to do the actual cleanup .